### DISCUSSION TITLE
systemctl

### DISCUSSION THREAD

**Discussion by ahbanks** [lichess-org/fishnet#177] [General]
This is likely old hat for most and likely done "in the dumbest way possible", but in case anyone else wants to run fishnet under systemctl, here's what I did on CentOS Stream:

Moved fishnet-x86_64-unknown-linux-gnu executable to /usr/local/bin
Moved fishnet.ini to /etc

Created /usr/local/bin/fishnet.sh:

```
#!/bin/bash
/usr/local/bin/fishnet-x86_64-unknown-linux-gnu --conf /etc/fishnet.ini
```

Created /etc/systemd/system/fishnet.service:

```
# Executable binary goes in /usr/local/bin
# fishnet.ini goes in /etc/fishnet.ini
[Unit]
Description=Lichess.org fishnet client
[Service]
User=<user you want to use>
WorkingDirectory=/usr/local/bin
ExecStart=/usr/local/bin/fishnet.sh
Restart=always
[Install]
WantedBy=multi-user.target
```

After that, I enabled the service:
```
systemctl enable --now fishnet.service
systemctl start fishnet
systemctl status fishnet
```

If you want to look at the log:
`sudo tail -f /var/log/messages`

**Comment by niklasf** [comment]
Thanks for sharing. Run

```
/usr/local/bin/fishnet-x86_64-unknown-linux-gnu --conf /etc/fishnet.ini systemd
```

to generate an even better service file. No helper script like `fishnet.sh` should be required.

**Reply by ahbanks** [reply]
Cool, I learned something today!  :-)

Will try that when I'm back home and update the post.

**Reply by ahbanks** [reply]
Quick update, I did the following:

```
mv fishnet.service fishnet.service.old

cd /etc/systemd/system

/usr/local/bin/fishnet-x86_64-unknown-linux-gnu --conf /etc/fishnet.ini systemd >> fishnet.service

systemctl daemon-reload

systemctl start fishnet

systemctl status fishnet
```

Seeing a running status, so this looks good!

**Reply by ahbanks** [reply]
Is fishnet slower under systemctl?  Nothing quantitative here, just me looking at the live logs.  I seem to see a lot more activity with fewer pauses when running natively instead of through systemctl.  

What would be the best metric to evaluate this?  Time between the batch messages and compute batch/min?

**Reply by niklasf** [reply]
Possibly slightly due to the `Nice=5`, if your server is doing other work. Other than that, no difference is expected.

**Reply by ahbanks** [reply]
Ok.  Aside from normal user interaction (cockpit, ssh, etc), nothing else is running on this system.

**Reply by preferredindifferent** [reply]
Hi there,
I know this is somewhat necromancing the thread - two quick questions: "User you want to use", is that the local machine user? And secondly, where would one put the private key to use?
Thanks in advance, xDSx

(I am looking to adapt this solution to launchctl on macOS, as I would like to run it in the background on a work machine where I am not in all the time)

**Reply by niklasf** [reply]
As defense in depth, it could also be a new user that doesn't have any privileges.

The private key can be in a `fishnet.ini` referenced using `--conf /path/to/fishnet.ini`:

```
[Fishnet]
key = ***
```

There's also `--key ***`, but note that the key might then appear in the process list.

**Reply by preferredindifferent** [reply]
thanks for the quick response, I will try to implement this soon! (I'm new to launchctl and their plist sytsem)
